contracts/automation/andromeda-oracle/src/contract.rs

```diff
@@ -0,0 +1,365 @@
+        info,
+        BaseInstantiateMsg {
+            ado_type: "oracle".to_string(),
+            ado_version: CONTRACT_VERSION.to_string(),
+            operators: None,
+            modules: None,
+            primitive_contract: None,
+        },
+    )
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn reply(_deps: DepsMut, _env: Env, _msg: Reply) -> Result<Response, ContractError> {
+    Ok(Response::default())
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    let contract = ADOContract::default();
+    match msg {
+        ExecuteMsg::AndrReceive(msg) => contract.execute(deps, env, info, msg, execute),
+    }
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn migrate(deps: DepsMut, _env: Env, _msg: MigrateMsg) -> Result<Response, ContractError> {
+    // New version
+    let version: Version = CONTRACT_VERSION.parse().map_err(from_semver)?;
+
+    // Old version
+    let stored = get_contract_version(deps.storage)?;
+    let storage_version: Version = stored.version.parse().map_err(from_semver)?;
+
+    let contract = ADOContract::default();
+
+    ensure!(
+        stored.contract == CONTRACT_NAME,
+        ContractError::CannotMigrate {
+            previous_contract: stored.contract,
+        }
+    );
+
+    // New version has to be newer/greater than the old version
+    ensure!(
+        storage_version < version,
+        ContractError::CannotMigrate {
+            previous_contract: stored.version,
+        }
+    );
+
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    // Update the ADOContract's version
+    contract.execute_update_version(deps)?;
+
+    Ok(Response::default())
+}
+
+fn from_semver(err: semver::Error) -> StdError {
+    StdError::generic_err(format!("Semver: {}", err))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, env: Env, msg: QueryMsg) -> Result<Binary, ContractError> {
+    match msg {
+        QueryMsg::AndrQuery(msg) => ADOContract::default().query(deps, env, msg, query),
+        QueryMsg::CurrentTarget {} => encode_binary(&query_current_target(deps)?),
+        QueryMsg::Target {} => encode_binary(&query_target(deps)?),
+        QueryMsg::StoredMessage {} => encode_binary(&query_stored_message(deps)?),
+    }
+}
+
+fn query_stored_message(deps: Deps) -> Result<String, ContractError> {
+    let message = QUERY_MSG.load(deps.storage)?;
+    Ok(message)
+}
+
+fn query_current_target(deps: Deps) -> Result<String, ContractError> {
+    let address = TARGET_ADO_ADDRESS.load(deps.storage)?;
+    Ok(address)
+}
+
+fn query_target(deps: Deps) -> Result<String, ContractError> {
+    let contract_addr = TARGET_ADO_ADDRESS.load(deps.storage)?;
+    let stored_msg = QUERY_MSG.load(deps.storage)?;
+
+    let decoded_string = base64::decode(stored_msg).unwrap();
```

Comment:
See comment in `oracle.rs` about typing, should make this simpler. Is there a reason it has to be a string as opposed to `Binary`?